sympy/core/equation.py

```diff
@@ -0,0 +1,378 @@
+    #####
+    def expand(self, *args, **kwargs):
+        return Equation(self.lhs.expand(*args, **kwargs), self.rhs.expand(
+            *args, **kwargs), check=False)
+
+    def simplify(self, *args, **kwargs):
+        return self._eval_simplify(*args, **kwargs)
+
+    def _eval_simplify(self, *args, **kwargs):
+        return Equation(self.lhs.simplify(*args, **kwargs), self.rhs.simplify(
+            *args, **kwargs), check=False)
+
+    def _eval_factor(self, *args, **kwargs):
+        # TODO: cancel out factors common to both sides.
+        return Equation(self.lhs.factor(*args, **kwargs), self.rhs.factor(
+            *args, **kwargs), check=False)
+
+    def factor(self, *args, **kwargs):
+        return self._eval_factor(*args, **kwargs)
+
+    def _eval_collect(self, *args, **kwargs):
+        from sympy.simplify.radsimp import collect
+        return Equation(collect(self.lhs, *args, **kwargs),
+                        collect(self.rhs, *args, **kwargs), check=False)
+
+    def collect(self, *args, **kwargs):
+        return self._eval_collect(*args, **kwargs)
+
+    def evalf(self, *args, **kwargs):
+        return Equation(self.lhs.evalf(*args, **kwargs),
+                        self.rhs.evalf(*args, **kwargs), check=False)
+
+    def _eval_derivative(self, *args, **kwargs):
+        # TODO Find why diff and Derivative do not appear to pass through
+        #  kwargs to this. Since we cannot set evaluation of lhs manually
+        #  try to be intelligent about when to do it.
+        from sympy.core.function import Derivative
+        eval_lhs = False
+        if not (isinstance(self.lhs, Derivative)):
+            for sym in args:
+                if sym in self.lhs.free_symbols and not (
+                _sympify(sym).is_number):
+                    eval_lhs = True
+        return Equation(self.lhs.diff(*args, **kwargs, evaluate=eval_lhs),
+                        self.rhs.diff(*args, **kwargs), check=False)
+
+    def integ(self, *args, **kwargs):
+        """
+        This function is a convenience function that returns a new equation
+        consisting of an unevaluated integral of the lhs as the new lhs and
+        the result of integrating the rhs as the new rhs.
+        """
+        from sympy.integrals.integrals import Integral
+        return Equation(Integral(self.lhs, *args, **kwargs),
+                        self.rhs.integrate(*args, **kwargs), check=False)
+
+    def _eval_Integral(self, *args, **kwargs):
+        side = kwargs.pop('side', None)  # Could not seem to pass values for
+        # `evaluate` through to here.
+        if side is None:
+            raise ValueError('You must specify `side="lhs"` or `side="rhs"` '
+                             'when integrating an Equation')
+        else:
+            try:
+                return (getattr(self, side).integrate(*args, **kwargs))
+            except AttributeError:
+                raise AttributeError('`side` must equal "lhs" or "rhs".')
```

Comment:
I suggest using `__getattr__` instead of manually defining these methods. With careful design, it saved a lot of work for me.

Reply:
I am not sure what you are suggesting. Most of the code surrounding `return (getattr(self, side).integrate(*args, **kwargs))` is to allow and force the use of the `side` keyword as you really do need to be thoughtful about how you want the integration applied to each side.

Reply:
Here is the code from my work, modified to be compatible with your design. Some `Basic` methods such as `rewrite` still has to be manually defined, but this can handle non-`Basic` methods like `diff` or `integrate`.

```python
from functools import partial

class Equation(Basic):
    ...
    def apply_attr(self, name, side="both"):
        """
        Get the attribute of the arguments and build Equation
        with the result.

        Parameters
        ==========

        name : str
            Name of the attribute

        side : "both", "lhs" or "rhs", optional
            Specify which term the method will be applied. Default is
            "both".

        """
        lhs, rhs = self.args
        if side in ("both", "lhs"):
            lhs = getattr(self.lhs, name)
        if side in ("both", "rhs"):
            rhs = getattr(self.rhs, name)
        return self.func(lhs, rhs)

    def apply_method(self, name, *args, side="both", **kwargs):
        """
        Apply the method on the arguments and build Equation
        with the result.

        Parameters
        ==========

        name : str
            Name of the method

        args : arguments passed to method

        side : "both", "lhs" or "rhs", optional
            Specify which term the method will be applied. Default is
            "both".

        kwargs : keyword arguments passed to method

        """
        lhs, rhs = self.args
        if side in ("both", "lhs"):
            lhs = getattr(self.lhs, name)(*args, **kwargs)
        if side in ("both", "rhs"):
            rhs = getattr(self.rhs, name)(*args, **kwargs)
        return self.func(lhs, rhs)

    def __getattr__(self, name):
        try:
            return self.__getattribute__(name)
        except AttributeError:
            lhs_attr = getattr(self.lhs, name)
            rhs_attr = getattr(self.rhs, name)
            if not (callable(lhs_attr) and callable(rhs_attr)):
                return partial(self.apply_attr, name)
            elif (callable(lhs_attr) and callable(rhs_attr)):
                return partial(self.apply_method, name)
            else:
                raise TypeError("Inconsistent methods are called on each side.")
```

Reply:
I like the apparent efficiency of this. But I am missing something. What does the `func` in this statement `return self.func(lhs, rhs)` refer to?

Reply:
A problem I have using this implementation is that `name` may not be defined in the namespace we are operating in. It may be implemented at a higher level. If the method is passed as an object then anything the interpreter knows about works. I will see if I can adjust these ideas to include that flexibility which my present less elegant method has.

Reply:
Incorporated while accommodating requests for alternative syntaxes. See [this comment](https://github.com/sympy/sympy/pull/19479#issuecomment-637859457) and discussions above it. See also [commit 57ef7ab](https://github.com/sympy/sympy/pull/19479/commits/57ef7ab1cd42bebd2bbe5ed8b90c6c413b192b44) and [commit e9044ac](https://github.com/sympy/sympy/pull/19479/commits/e9044ac7fd52f35814749a0af75ddea49a499145).